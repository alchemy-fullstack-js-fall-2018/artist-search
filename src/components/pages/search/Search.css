.resultList {
  list-style: none;
  margin: 16px 25%;
  text-align: left;
}
