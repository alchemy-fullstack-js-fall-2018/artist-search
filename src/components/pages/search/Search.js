import React, { Component, Fragment } from 'react';
import PropTypes from 'prop-types';
import { withRouter } from 'react-router-dom';

import Result from './result/Result';
import Pager from '../../lib/pager/Pager';

import styles from './Search.css';

import queryString from 'querystring';
import { getArtists } from '../../../services/musicSearch';
import { ROUTES } from '../../../routes/routes';

class Search extends Component {
  static propTypes = {
    location: PropTypes.object.isRequired,
    history: PropTypes.object.isRequired
  };

  state = {
    totalPages: null,
    artists: []
  };

  componentDidMount() {
    this.doSearch();
  }

  componentDidUpdate(prevProps, prevState) {
    if(prevProps.location.search !== this.props.location.search) {
      this.doSearch();
    }
  }

  doSearch = () => {
    const { q, page } = queryString.parse(this.props.location.search.slice(1));
    if(!q) return;

    getArtists(q, page)
      .then(({ artists, totalPages }) => {
        this.setState({ artists, totalPages });
      });
  };

  updateHistory = (options) => {
    const { page: urlPage, q: urlSearchTerm } = queryString.parse(this.props.location.search.slice(1));
    const q = options.q || urlSearchTerm;
    const page = options.page || urlPage;
    const searchQuery = queryString.stringify({ q, page });
    this.props.history.push({
      pathName: ROUTES.SEARCH.path,
      search: searchQuery
    });
  };


  updateSearch = event => {
    event.preventDefault();
    const q = document.getElementById('q').value;
    this.updateHistory({ q, page: 0 });
  };

  updatePage = page => {
    this.updateHistory({ page });
    this.doSearch();
  };

  render() {
    const { totalPages } = this.state;

    const { page } = queryString.parse(this.props.location.search.slice(1));
    const pageNum = parseInt(page);

    const results = this.state.artists.map(artist => {
      return (
        <li key={artist.id}>
          <Result
            id={artist.id}
            name={artist.name}
            description={artist.description}
          />
        </li>
      );
    });

    return (
      <Fragment>
        <h2>Search the Music Database</h2>
        <form
          onSubmit={this.updateSearch}
        >
          <input id="q" type="text" />
          <button type="submit">Search</button>
        </form>
        <ul
          className={styles.resultList}
        >
          {results}
        </ul>
        {
          totalPages &&
          <Pager
            currentPage={pageNum}
            totalPages={totalPages}
            updatePage={this.updatePage}
          />
        }
      </Fragment>
    );
  }
}

export default withRouter(Search);
