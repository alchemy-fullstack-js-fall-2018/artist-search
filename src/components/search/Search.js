import React, { Component, Fragment } from 'react';
import PropTypes from 'prop-types';
import queryString from 'querystring';
import { withRouter, Link } from 'react-router-dom';
import { getArtists } from '../../services/musicSearch';
import { ROUTES } from '../../routes/routes';
import Artist from '../artist/Artist';
import styles from './search.css';
import Paging from '../paging/Paging';


class Search extends Component {
  static propTypes = {
    location: PropTypes.object.isRequired,
    history: PropTypes.object.isRequired
  };

  state = {
    totalPages: null,
    artists: []
  };

  doSearch = () => {
    const { searchTerm, page } = queryString.parse(this.props.location.search.slice(1));
    if(!searchTerm) return;

    getArtists(searchTerm, page)
      .then(({ pages, artists }) => this.setState({ artists, totalPages: pages }));
  };

  componentDidMount() {
    this.doSearch();
  }

  handlePageUpdate = page => {
    this.updateHistory({ page });
  };

  componentDidUpdate(prevProps, prevState) {

    if(prevProps.location.search !== this.props.location.search) {
      this.doSearch();
    }
  }

  updateSearchTerm = event => {
    event.preventDefault();
    const searchTerm = document.getElementById('searchTerm').value;
    this.updateHistory({ searchTerm, page: 1 });
  };

updateHistory = search => {
  const { searchTerm: currentSearchTerm, page: currentPage } = queryString.parse(this.props.location.search.slice(1));
  const searchTerm = search.searchTerm || currentSearchTerm;
  const page = search.page || currentPage;
  const searchQuery = queryString.stringify({ searchTerm, page });
    this.props.history.push({
      pathName: ROUTES.SEARCH.path,
      search: searchQuery
    });
}

  render() {
    const { totalPages  } = this.state;
    const { page } = queryString.parse(this.props.location.search.slice(1));
    const artistsComponent = this.state.artists.map(artist => {
      return (
        <li key={artist.id}>
          <Artist id={artist.id}
            name={artist.name}
            description={artist.description} />
        </li>
      );
    });

    return (
      <Fragment>

        <Paging currentPage={parseInt(page)}
          totalPages={totalPages}
          updatePage={this.handlePageUpdate} />
        <Link to={ROUTES.LANDING.linkTo()}>Link!</Link>

        <form onSubmit={this.updateSearchTerm}>
          <input id="searchTerm" type="text" />
          <button type="submit">Search</button>
        </form>
        <ul className={styles.search}>
          {artistsComponent}
        </ul>
      </Fragment >
    );
  }
}

export default withRouter(Search);
