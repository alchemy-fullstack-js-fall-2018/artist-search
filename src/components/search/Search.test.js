import React from 'react';
import Search from './Search';
import { shallow } from 'enzyme';

describe('search', () => {
    let wrapper;
    
    beforeEach(() => {
        wrapper = shallow(<Search />);
    });

    it('renders a search component', () => {
        expect(wrapper).toMatchSnapshot();
    });
});
