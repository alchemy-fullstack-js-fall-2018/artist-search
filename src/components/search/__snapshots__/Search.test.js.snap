// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`search renders a search component 1`] = `
<Fragment>
  <h2>
    Artist Search
  </h2>
  <div>
    <form>
      <label>
         Search for Artists By Name:
        <input
          name="artistName"
          onChange={[Function]}
          type="text"
          value=""
        />
      </label>
    </form>
  </div>
</Fragment>
`;
